Crypto/OTP.hs

```diff
@@ -0,0 +1,168 @@
+    -- or Nothing if the submitted OTP values didn't match anywhere within the window
+resynchronize h d s k c (p1, extras) = do
+    offBy <- fmap fromIntegral (elemIndex p1 range)
+    checkExtraOtps (c + offBy + 1) extras
+  where
+    checkExtraOtps ctr [] = Just ctr
+    checkExtraOtps ctr (p:ps)
+        | hotp h d k ctr /= p = Nothing
+        | otherwise           = checkExtraOtps (ctr + 1) ps
+
+    range = map (hotp h d k)[c..c + fromIntegral s]
+
+digitsPower :: OTPDigits -> Word32
+digitsPower OTP4 = 10000
+digitsPower OTP5 = 100000
+digitsPower OTP6 = 1000000
+digitsPower OTP7 = 10000000
+digitsPower OTP8 = 100000000
+digitsPower OTP9 = 1000000000
+
+
+data TOTPParams h = TP !h !Word64 !Word16 !OTPDigits !ClockSkew
+
+data ClockSkew = NoSkew | OneStep | TwoSteps | ThreeSteps | FourSteps deriving (Enum)
+
+-- | The default TOTP configuration.
+defaultTOTPParams :: TOTPParams SHA1
+defaultTOTPParams = TP SHA1 0 30 OTP6 TwoSteps
+
+-- | Create a TOTP configuration with customized parameters.
+mkTOTPParams :: (HashAlgorithm hash)
+    => hash
+    -> Word64
+    -- ^ The T0 parameter in seconds. This is the Unix time from which to start
+    -- counting steps (default 0). Must be before the current time.
+    -> Word16
+    -- ^ The time step parameter X in seconds (default 30, maximum allowed 300)
+    -> OTPDigits
+    -- ^ Number of required digits in the OTP (default 6)
+    -> ClockSkew
+    -- ^ The number of time steps to check either side of the current value
+    -- to allow for clock skew between client and server and or delay in
+    -- submitting the value. The default is two time steps.
+    -> Either String (TOTPParams hash)
+mkTOTPParams h t0 x d skew = do
+    unless (x > 0) (Left "Time step must be greater than zero")
+    unless (x <= 300) (Left "Time step cannot be greater than 300 seconds")
+    return (TP h t0 x d skew)
+
+-- | Calculate a totp value for the given time.
+totp :: (HashAlgorithm hash, ByteArrayAccess key)
+    => TOTPParams hash
+    -> key
+    -- ^ The shared secret
+    -> POSIXTime
+    -- ^ The time for which the OTP should be calculated.
+    -- This is usually the current time as returned by @Data.Time.Clock.POSIX.getPOSIXTime@
+    -> Word32
+    -- ^ The OTP value
+totp (TP h t0 x d _) k now = hotp h d k (timeToCounter now t0 x)
+
+-- | Check a supplied TOTP value is valid for the given time,
+-- within the window defined by the skew parameter.
+totpVerify :: (HashAlgorithm hash, ByteArrayAccess key)
+    => TOTPParams hash
+    -> key
+    -> POSIXTime
+    -> Word32
+    -> Bool
+totpVerify (TP h t0 x d skew) k now otp = otp `elem` map (hotp h d k) (range window [])
+  where
+    t = timeToCounter now t0 x
+    window = fromIntegral (fromEnum skew)
+    range 0 acc = t : acc
+    range n acc = range (n-1) ((t-n) : (t+n) : acc)
+
+timeToCounter :: POSIXTime -> Word64 -> Word16 -> Word64
+timeToCounter now t0 x = floor ((now - fromIntegral t0) / fromIntegral x)
```

Comment:
They could (I would :-)) use `timeCurrent` from `hourglass` as well. I'm not sure it's a good idea to have a IO callback, as this would make the function impure, and un-replayable (if you want to test known values or passed/future values).